### PR TITLE
planner: let projection can push down to TiFlash when mpp is not enforced.

### PR DESCRIPTION
**Pull request by LittleFall**
### What problem does this PR solve?

Problem Summary: https://github.com/pingcap/tidb/pull/25450 support projection push down to TiFlash, but it only takes effect when mpp mode is enforced. This changes the semantics of `tidb_enforce_mpp`: It will introduce new optimization rules instead of just ignoring the optimizer cost estimate.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed: make this optimization rule takes effect in the normal situation.

### Check List 

Tests 

- [x] Integration test

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory

It may bring more optimizer search space and branches, but in TP business, it will be blocked by the following check `p.canPushToCop`.

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

- Support projection push down to TiFlash in mpp mode to speed up queries.

### CODE REVIEW

File: bindinfo/bind_test.go [hunk 1]
```diff
@@ -1635,8 +1635,8 @@ func (s *testSuite) TestNotEvolvePlanForReadStorageHint(c *C) {
 	s.cleanBindingEnv(tk)
 	tk.MustExec("use test")
 	tk.MustExec("drop table if exists t")
-	tk.MustExec("create table t(a int, b int, index idx_a(a), index idx_b(b))")
-	tk.MustExec("insert into t values (1,1), (2,2), (3,3), (4,4), (5,5), (6,6), (7,7), (8,8), (9,9), (10,10)")
+	tk.MustExec("create table t(a int, index idx_a(a))")
```
Comment (hanfei1991) [inline bindinfo/bind_test.go:1638]: What happened to these test?

File: bindinfo/bind_test.go [hunk 2]
```diff
@@ -1653,22 +1653,22 @@ func (s *testSuite) TestNotEvolvePlanForReadStorageHint(c *C) {
...
+	rows := tk.MustQuery("explain select * from t where a >= 11").Rows()
 	c.Assert(fmt.Sprintf("%v", rows[len(rows)-1][2]), Equals, "cop[tikv]")
 
-	tk.MustExec("create global binding for select * from t where a >= 1 and b >= 1 using select /*+ read_from_storage(tiflash[t]) */ * from t where a >= 1 and b >= 1")
+	tk.MustExec("create global binding for select * from t where a >= 1 using select /*+ read_from_storage(tiflash[t]) */ * from t where a >= 1")
 	tk.MustExec("set @@tidb_evolve_plan_baselines=1")
 
 	// Even if index of TiKV has lower cost, it chooses TiFlash.
-	rows = tk.MustQuery("explain select * from t where a >= 11 and b >= 11").Rows()
+	rows = tk.MustQuery("explain select * from t where a >= 11").Rows()
```
Comment (lzmhhh123) [inline bindinfo/bind_test.go:1664]: Why change the binding tests?

**Comment by LittleFall** [comment]
after #25450 and this pr is patched, the

**Comment by LittleFall** [comment]
> after #25450 and this pr is patched, the cost of some plan will be changed (The cost of plan on tiflash is reduced), �so some test is broken.

For example, the bind test. In this test, it is assumed that it will select tikv. In fact, it selected tiflash. In order to ensure that it reaches tikv, I modified the test sql. 
@lzmhhh123 @hanfei1991

**Comment by LittleFall** [comment]
another things is after this patch, the estCost in `explain format='verbose'` seems not accurate. 

```sql
drop table t;
create table t(a int, index idx(a));
alter table t set tiflash replica 1;
analyze table t;
explain format='verbose' select /*+ read_from_storage(tikv[t]) */ * from t where a>=11; -- 11473
explain format='verbose' select * from t where a>=11;         -- 12098
explain format='verbose' select /*+ read_from_storage(tiflash[t]) */ * from t where a>=11; -- 12098
```

**Comment by LittleFall** [comment]
I will close this pr because the influence on optimizer cost model can't be evaluated.

**Comment by LittleFall** [comment]
I will close this pr because the influence on optimizer cost model can't be evaluated.

**Pull request closed without merging**